In [1]:
import pandas as pd
import numpy as np

In [2]:
pew_df = pd.read_csv(r'C:\Users\taewoo\PycharmProjects\data_professionals\Projects\Datasets\pew.csv')

### Tidying Data #1

In [3]:
print(pew_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18 entries, 0 to 17
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   religion            18 non-null     object
 1   <$10k               18 non-null     int64 
 2   $10-20k             18 non-null     int64 
 3   $20-30k             18 non-null     int64 
 4   $30-40k             18 non-null     int64 
 5   $40-50k             18 non-null     int64 
 6   $50-75k             18 non-null     int64 
 7   $75-100k            18 non-null     int64 
 8   $100-150k           18 non-null     int64 
 9   >150k               18 non-null     int64 
 10  Don't know/refused  18 non-null     int64 
dtypes: int64(10), object(1)
memory usage: 1.7+ KB
None


In [4]:
pew_df.head()

religion  <$10k  $10-20k  $20-30k  $30-40k  $40-50k  $50-75k  \
0            Agnostic     27       34       60       81       76      137   
1             Atheist     12       27       37       52       35       70   
2            Buddhist     27       21       30       34       33       58   
3            Catholic    418      617      732      670      638     1116   
4  Don’t know/refused     15       14       15       11       10       35   

   $75-100k  $100-150k  >150k  Don't know/refused  
0       122        109     84                  96  
1        73         59     74                  76  
2        62         39     53                  54  
3       949        792    633                1489  
4        21         17     18                 116

Notice the the design of the table is inconvenient to navigate. I want to *unpivot* the table.

This is when `melt()` is used.

In [5]:
# id_vars are identifier columns
melted_pew = pew_df.melt(id_vars='religion', 
                         var_name='Income',
                         value_name='Count')

melted_pew.head(10)

religion Income  Count
0                 Agnostic  <$10k     27
1                  Atheist  <$10k     12
2                 Buddhist  <$10k     27
3                 Catholic  <$10k    418
4       Don’t know/refused  <$10k     15
5         Evangelical Prot  <$10k    575
6                    Hindu  <$10k      1
7  Historically Black Prot  <$10k    228
8        Jehovah's Witness  <$10k     20
9                   Jewish  <$10k     19

In [6]:
print(pew_df.shape)
print(melted_pew.shape)

(18, 11)
(180, 3)


Essentially, it is transforming wide format data into narrow format data by unpivoting values

### Tidying Data #2

In [7]:
billboard = pd.read_csv(r'C:\Users\taewoo\PycharmProjects\data_professionals\Projects\Datasets\billboard.csv')

In [8]:
print(billboard.shape)
print(billboard.columns)

(317, 81)
Index(['year', 'artist', 'track', 'time', 'date.entered', 'wk1', 'wk2', 'wk3',
       'wk4', 'wk5', 'wk6', 'wk7', 'wk8', 'wk9', 'wk10', 'wk11', 'wk12',
       'wk13', 'wk14', 'wk15', 'wk16', 'wk17', 'wk18', 'wk19', 'wk20', 'wk21',
       'wk22', 'wk23', 'wk24', 'wk25', 'wk26', 'wk27', 'wk28', 'wk29', 'wk30',
       'wk31', 'wk32', 'wk33', 'wk34', 'wk35', 'wk36', 'wk37', 'wk38', 'wk39',
       'wk40', 'wk41', 'wk42', 'wk43', 'wk44', 'wk45', 'wk46', 'wk47', 'wk48',
       'wk49', 'wk50', 'wk51', 'wk52', 'wk53', 'wk54', 'wk55', 'wk56', 'wk57',
       'wk58', 'wk59', 'wk60', 'wk61', 'wk62', 'wk63', 'wk64', 'wk65', 'wk66',
       'wk67', 'wk68', 'wk69', 'wk70', 'wk71', 'wk72', 'wk73', 'wk74', 'wk75',
       'wk76'],
      dtype='object')


In [9]:
melted_billboard = billboard.melt(id_vars=['year', 'artist', 'track', 'time', 'date.entered'],
                                  var_name='Week',
                                  value_name='Rating')
melted_billboard.head()

year        artist                    track  time date.entered Week  Rating
0  2000         2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk1    87.0
1  2000       2Ge+her  The Hardest Part Of ...  3:15   2000-09-02  wk1    91.0
2  2000  3 Doors Down               Kryptonite  3:53   2000-04-08  wk1    81.0
3  2000  3 Doors Down                    Loser  4:24   2000-10-21  wk1    76.0
4  2000      504 Boyz            Wobble Wobble  3:35   2000-04-15  wk1    57.0

Let's pick an artist and a track to see if everything is formatted properly

In [10]:
pd.set_option('display.max_row', 76)
melted_billboard.loc[(melted_billboard['artist'] == '3 Doors Down')&
                     (melted_billboard['track'] == 'Loser')]

year        artist  track  time date.entered  Week  Rating
3      2000  3 Doors Down  Loser  4:24   2000-10-21   wk1    76.0
320    2000  3 Doors Down  Loser  4:24   2000-10-21   wk2    76.0
637    2000  3 Doors Down  Loser  4:24   2000-10-21   wk3    72.0
954    2000  3 Doors Down  Loser  4:24   2000-10-21   wk4    69.0
1271   2000  3 Doors Down  Loser  4:24   2000-10-21   wk5    67.0
1588   2000  3 Doors Down  Loser  4:24   2000-10-21   wk6    65.0
1905   2000  3 Doors Down  Loser  4:24   2000-10-21   wk7    55.0
2222   2000  3 Doors Down  Loser  4:24   2000-10-21   wk8    59.0
2539   2000  3 Doors Down  Loser  4:24   2000-10-21   wk9    62.0
2856   2000  3 Doors Down  Loser  4:24   2000-10-21  wk10    61.0
3173   2000  3 Doors Down  Loser  4:24   2000-10-21  wk11    61.0
3490   2000  3 Doors Down  Loser  4:24   2000-10-21  wk12    59.0
3807   2000  3 Doors Down  Loser  4:24   2000-10-21  wk13    61.0
4124   2000  3 Doors Down  Loser  4:24   2000-10-21  wk14    66.0
4441   2000  3 Doors Down  Loser  4:24   2000-10-21  wk15    72.0
4758   2000  3 Doors Down  Loser  4:24   2000-10-21  wk16    76.0
5075   2000  3 Doors Down  Loser  4:24   2000-10-21  wk17    75.0
5392   2000  3 Doors Down  Loser  4:24   2000-10-21  wk18    67.0
5709   2000  3 Doors Down  Loser  4:24   2000-10-21  wk19    73.0
6026   2000  3 Doors Down  Loser  4:24   2000-10-21  wk20    70.0
6343   2000  3 Doors Down  Loser  4:24   2000-10-21  wk21     NaN
6660   2000  3 Doors Down  Loser  4:24   2000-10-21  wk22     NaN
6977   2000  3 Doors Down  Loser  4:24   2000-10-21  wk23     NaN
7294   2000  3 Doors Down  Loser  4:24   2000-10-21  wk24     NaN
7611   2000  3 Doors Down  Loser  4:24   2000-10-21  wk25     NaN
7928   2000  3 Doors Down  Loser  4:24   2000-10-21  wk26     NaN
8245   2000  3 Doors Down  Loser  4:24   2000-10-21  wk27     NaN
8562   2000  3 Doors Down  Loser  4:24   2000-10-21  wk28     NaN
8879   2000  3 Doors Down  Loser  4:24   2000-10-21  wk29     NaN
9196   2000  3 Doors Down  Loser  4:24   2000-10-21  wk30     NaN
9513   2000  3 Doors Down  Loser  4:24   2000-10-21  wk31     NaN
9830   2000  3 Doors Down  Loser  4:24   2000-10-21  wk32     NaN
10147  2000  3 Doors Down  Loser  4:24   2000-10-21  wk33     NaN
10464  2000  3 Doors Down  Loser  4:24   2000-10-21  wk34     NaN
10781  2000  3 Doors Down  Loser  4:24   2000-10-21  wk35     NaN
11098  2000  3 Doors Down  Loser  4:24   2000-10-21  wk36     NaN
11415  2000  3 Doors Down  Loser  4:24   2000-10-21  wk37     NaN
11732  2000  3 Doors Down  Loser  4:24   2000-10-21  wk38     NaN
12049  2000  3 Doors Down  Loser  4:24   2000-10-21  wk39     NaN
12366  2000  3 Doors Down  Loser  4:24   2000-10-21  wk40     NaN
12683  2000  3 Doors Down  Loser  4:24   2000-10-21  wk41     NaN
13000  2000  3 Doors Down  Loser  4:24   2000-10-21  wk42     NaN
13317  2000  3 Doors Down  Loser  4:24   2000-10-21  wk43     NaN
13634  2000  3 Doors Down  Loser  4:24   2000-10-21  wk44     NaN
13951  2000  3 Doors Down  Loser  4:24   2000-10-21  wk45     NaN
14268  2000  3 Doors Down  Loser  4:24   2000-10-21  wk46     NaN
14585  2000  3 Doors Down  Loser  4:24   2000-10-21  wk47     NaN
14902  2000  3 Doors Down  Loser  4:24   2000-10-21  wk48     NaN
15219  2000  3 Doors Down  Loser  4:24   2000-10-21  wk49     NaN
15536  2000  3 Doors Down  Loser  4:24   2000-10-21  wk50     NaN
15853  2000  3 Doors Down  Loser  4:24   2000-10-21  wk51     NaN
16170  2000  3 Doors Down  Loser  4:24   2000-10-21  wk52     NaN
16487  2000  3 Doors Down  Loser  4:24   2000-10-21  wk53     NaN
16804  2000  3 Doors Down  Loser  4:24   2000-10-21  wk54     NaN
17121  2000  3 Doors Down  Loser  4:24   2000-10-21  wk55     NaN
17438  2000  3 Doors Down  Loser  4:24   2000-10-21  wk56     NaN
17755  2000  3 Doors Down  Loser  4:24   2000-10-21  wk57     NaN
18072  2000  3 Doors Down  Loser  4:24   2000-10-21  wk58     NaN
18389  2000  3 Doors Down  Loser  4:24   2000-10-21  wk59     NaN
18706  2000  3 Doors Down  Loser  4:24   2000-1

In [11]:
print(melted_billboard.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24092 entries, 0 to 24091
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   year          24092 non-null  int64  
 1   artist        24092 non-null  object 
 2   track         24092 non-null  object 
 3   time          24092 non-null  object 
 4   date.entered  24092 non-null  object 
 5   Week          24092 non-null  object 
 6   Rating        5307 non-null   float64
dtypes: float64(1), int64(1), object(5)
memory usage: 1.3+ MB
None


Noticeable changes 
* Dramatic change of the shape (317, 81) -> (24092, 7)
* Memory usage went up from `200.7+ KB` to `1.3+ MB`
* There are numerous rows with nontype value for the `Rating` column

If I were to unpivot a much bigger dataset, memory usage might be a problem.

To reduce it, let's only select the rows that hold values for the `Rating` column

In [12]:
melted_billboard = melted_billboard[pd.notnull(melted_billboard['Rating'])]

In [13]:
melted_billboard.loc[(melted_billboard['artist'] == '3 Doors Down')&
                     (melted_billboard['track'] == 'Loser')]

year        artist  track  time date.entered  Week  Rating
3     2000  3 Doors Down  Loser  4:24   2000-10-21   wk1    76.0
320   2000  3 Doors Down  Loser  4:24   2000-10-21   wk2    76.0
637   2000  3 Doors Down  Loser  4:24   2000-10-21   wk3    72.0
954   2000  3 Doors Down  Loser  4:24   2000-10-21   wk4    69.0
1271  2000  3 Doors Down  Loser  4:24   2000-10-21   wk5    67.0
1588  2000  3 Doors Down  Loser  4:24   2000-10-21   wk6    65.0
1905  2000  3 Doors Down  Loser  4:24   2000-10-21   wk7    55.0
2222  2000  3 Doors Down  Loser  4:24   2000-10-21   wk8    59.0
2539  2000  3 Doors Down  Loser  4:24   2000-10-21   wk9    62.0
2856  2000  3 Doors Down  Loser  4:24   2000-10-21  wk10    61.0
3173  2000  3 Doors Down  Loser  4:24   2000-10-21  wk11    61.0
3490  2000  3 Doors Down  Loser  4:24   2000-10-21  wk12    59.0
3807  2000  3 Doors Down  Loser  4:24   2000-10-21  wk13    61.0
4124  2000  3 Doors Down  Loser  4:24   2000-10-21  wk14    66.0
4441  2000  3 Doors Down  Loser  4:24   2000-10-21  wk15    72.0
4758  2000  3 Doors Down  Loser  4:24   2000-10-21  wk16    76.0
5075  2000  3 Doors Down  Loser  4:24   2000-10-21  wk17    75.0
5392  2000  3 Doors Down  Loser  4:24   2000-10-21  wk18    67.0
5709  2000  3 Doors Down  Loser  4:24   2000-10-21  wk19    73.0
6026  2000  3 Doors Down  Loser  4:24   2000-10-21  wk20    70.0

In [14]:
print(melted_billboard.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5307 entries, 0 to 20350
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   year          5307 non-null   int64  
 1   artist        5307 non-null   object 
 2   track         5307 non-null   object 
 3   time          5307 non-null   object 
 4   date.entered  5307 non-null   object 
 5   Week          5307 non-null   object 
 6   Rating        5307 non-null   float64
dtypes: float64(1), int64(1), object(5)
memory usage: 331.7+ KB
None


### Tidying Data #3

In [15]:
weather = pd.read_csv(r'C:\Users\taewoo\PycharmProjects\data_professionals\Projects\Datasets\weather.csv')

In [16]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 35 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       22 non-null     object 
 1   year     22 non-null     int64  
 2   month    22 non-null     int64  
 3   element  22 non-null     object 
 4   d1       2 non-null      float64
 5   d2       4 non-null      float64
 6   d3       4 non-null      float64
 7   d4       2 non-null      float64
 8   d5       8 non-null      float64
 9   d6       2 non-null      float64
 10  d7       2 non-null      float64
 11  d8       2 non-null      float64
 12  d9       0 non-null      float64
 13  d10      2 non-null      float64
 14  d11      2 non-null      float64
 15  d12      0 non-null      float64
 16  d13      2 non-null      float64
 17  d14      4 non-null      float64
 18  d15      2 non-null      float64
 19  d16      2 non-null      float64
 20  d17      2 non-null      float64
 21  d18      0 non-nul